File: 2020/Day12/Python/RainRisk-Part1.py
import sys
import collections
import itertools
import numpy as np

pos = [0, 0]
direction = 90

instructions = []
for line in open(f"{sys.path[0]}/../input.txt").read().splitlines():
    move = line[:1]
    amount = int(line[1:])
    instructions.append((move, amount))

for ins, amount in instructions:
    if ins == "F":
        ins = "NESW"[direction // 90]
    elif ins == "R":
        direction += amount
        direction %= 360
    elif ins == "L":
        direction -= amount
        direction %= 360
    if ins == "N":      pos[1] += amount
    elif ins == "E":    pos[0] += amount
    elif ins == "S":    pos[1] -= amount
    elif ins == "W":    pos[0] -= amount
        
print(abs(pos[0]) + abs(pos[1]))

File: 2020/Day12/Haskell/RainRisk-Part1.hs
import System.IO
import Control.Monad

printArray:: (Show a, Eq a) => [a] -> String
printArray (x:xs)
    | null xs = show x
    | otherwise = show x ++ "\n" ++ printArray xs

solve:: [(Char, Int)] -> Int -> (Int, Int) -> Int
solve [] _ (x, y) = abs x + abs y
solve ((ins, amt):xs) dir (x, y) 
    | ins == 'F' = solve (("NESW"!!(dir `div` 90), amt) : xs) dir (x, y)
    | ins == 'R' = solve xs ((dir + amt) `mod` 360) (x, y)
    | ins == 'L' = solve xs ((dir - amt) `mod` 360) (x, y)
    | ins == 'N' = solve xs dir (x, y+amt)
    | ins == 'E' = solve xs dir (x+amt, y)
    | ins == 'S' = solve xs dir (x, y-amt)
    | ins == 'W' = solve xs dir (x-amt, y)

main:: IO()
main = do
    handle <- openFile "../input.txt" ReadMode
    contents <- hGetContents handle
    let rows = lines contents
    let instructions = map(\x -> (head x, read (tail x)::Int)) rows
    print (solve instructions 90 (0, 0))
    hClose handle
